astroid/constraint.py

```diff
@@ -0,0 +1,135 @@
+# Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Classes representing different types of constraints on inference values."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TypeVar, Union
+
+from astroid import nodes, util
+
+_NameNodes = Union[nodes.AssignAttr, nodes.Attribute, nodes.AssignName, nodes.Name]
+_ConstraintT = TypeVar("_ConstraintT", bound="Constraint")
+
+
+class Constraint(ABC):
+    """Represents a single constraint on a variable."""
+
+    def __init__(self, node: nodes.NodeNG, negate: bool) -> None:
+        self.node = node
+        """The node that this constraint applies to."""
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```

Comment:
Since this is a `classmethod`, the type of the first argument should be a class type not an instance.
```suggestion
    @classmethod
    @abstractmethod
    def match(
        cls: type[_ConstraintT], node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
```

Reply:
Changed this to be `Self` which I think makes more sense here, let me know if I need to revert

Reply:
`Self` would be awesome. Unfortunately it isn't supported by mypy yet.
The PR has already been merged, but we need to wait for the next release (1.0.0).

Reply:
Yeah, because it was already merged I thought it would be okay to use. `pyright` understands it and I don't think we will turn on `mypy` in `astroid` before `1.0.0` gets released so this doesn't give us any additional work (instead it prevents some :smile: )